File: src/main/java/com/application/lamion/controller/api/MakeRequestController.java
package com.application.lamion.controller.api;

import com.application.lamion.exception.BadRequest;
import com.application.lamion.exception.NotFound;
import com.application.lamion.form.MakeRequestForm;
import com.application.lamion.model.Event;
import com.application.lamion.model.Request;
import com.application.lamion.service.AppService;
import com.application.lamion.service.EventService;
import com.application.lamion.service.RequestService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@CrossOrigin
@RestController
@RequestMapping("/api")
public class MakeRequestController {
    private AppService appService;
    private EventService eventService;
    private RequestService requestService;

    @Autowired
    public MakeRequestController(AppService appService, EventService eventService, RequestService requestService) {
        this.appService = appService;
        this.eventService = eventService;
        this.requestService = requestService;
    }

    @PostMapping("request")
    public ResponseEntity makeRequest(@RequestBody MakeRequestForm form) {
        if (form.isValid()) {
            if (appService.exists(form.appId)) {
                Event event = eventService.findByTitle(form.event);
                if (event == null) {
                    event = new Event();
                    event.setApplicationId(form.appId);
                    event.setTitle(form.event);
                    event = eventService.create(event);
                }

                Request request = new Request();
                request.setEventId(event.getId());
                request.setDevice(form.device);
                requestService.create(request);

                return ResponseEntity.status(204).build();
            }

            return ResponseEntity.status(404).body(new NotFound());
        }

        return ResponseEntity.badRequest().body(new BadRequest());
    }
}

File: server/src/main/java/com/application/lamion/controller/api/MakeRequestController.kt
package com.application.lamion.controller.api

import com.application.lamion.exception.BadRequest
import com.application.lamion.exception.NotFound
import com.application.lamion.form.MakeRequestForm
import com.application.lamion.model.Event
import com.application.lamion.model.Request
import com.application.lamion.service.AppService
import com.application.lamion.service.EventService
import com.application.lamion.service.RequestService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*

@CrossOrigin
@RestController
@RequestMapping("/api")
class MakeRequestController @Autowired constructor(
    private val appService: AppService,
    private val eventService: EventService,
    private val requestService: RequestService
) {
    @PostMapping("request")
    fun makeRequest(@RequestBody form: MakeRequestForm): ResponseEntity<*> {
        if (form.isValid) {
            if (appService.exists(form.appId.toLong())) {
                val event = eventService.findByTitleAndApplicationId(form.event, form.appId.toLong())
                    ?: Event(
                        applicationId = form.appId,
                        title = form.event!!,
                        date = Date(),
                    ).let {
                        eventService.create(it)
                    }

                val request = Request(
                    eventId = event.id,
                    device = form.device!!,
                    date = Date()
                )

                requestService.save(request)

                return ResponseEntity.status(204).build<Any>()
            }

            return ResponseEntity.status(404).body(NotFound())
        }

        return ResponseEntity.badRequest().body(BadRequest())
    }
}
